### PR TITLE
Piano+LibAudio+LibIPC: Realtime enqueueing, stage 1

### PR DESCRIPTION
**Pull request by kleinesfilmroellchen**
Chunk 7 of #16049

TL;DR: Without changing latency, Piano now runs without technical (i.e. measurable but not hearable) or hearable audio glitches. Before, it was running with strongly noticeable glitches, which would worsen once the system load increased. After this change, Piano's audio performance is almost independent of system load, remaining glitch-free as long as there are necessary CPU resources (which is <~8% of CPU time at the moment).

Note that this PR does the minimal amount of work required to improve the status quo in a first step, will be improved upon later, and does intentionally not contain changes to LibDSP.

These changes additionally required
- some minor changes to LibAudio so that we can access the correct APIs in the first place
- decoupling LibIPC from EventLoop for good (Ladybird might want to take advantage of this!)
- a lot of atomics and mutexes so that we're multithreading safely

It also includes a framerate decrease for Piano's UI because the drawing code is really slow and the high, unsynced framerate can cause problems for WindowServer.

### Performance measurements

These measurements were all done with a 2000 measurement moving average, as individual pipeline times jump around wildly.

Before this PR, the full pipeline of audio generation -> resampling -> sending it off to the queue takes around 1.5ms - 3.5ms under normal loads (three tracks and several notes at the same time on all tracks), and when the system load increases, this time frequently jumps up to 6ms and higher.

#### However - large caveat:

This does not take into account the background thread which is enqueueing audio to the audio server. This thread, under the same circumstances described above, is too slow at a rate of ~3Hz with some 10Hz bursts (based on the AudioServer complain messages in the console, available on a normal master build, try it yourself). **This means that no matter how few milliseconds the audio pipeline actually takes, the data processing between these three threads and two processes is too slow at least three times per second.** This is audible.

#### After this PR

First of all, there are no complaint messages from the AudioServer under described normal circumstances. By never, I mean not in the several hours I have tested this code. The only instances of complaint messages is with extremely high system load, which is to be expected and not a problem specific to Piano.

The same performance measurements, **this time even including the total audio enqueue time** (but other than that, measuring the same computations as before), result in an *extremely consistent* average of 1.8ms - 2.2ms. Statistically speaking, the mean has increased, but the variance has decreased much more.

I hope I have convinced you that (1) measuring performance across multithreading refactors is more subjective than you think and (2) trusting your ears (especially when the console messages agree with your ears) is something that is necessary for audio programming.

I hope I have further convinced you that this new architecture is not, in fact, more complicated than the old one. I consider it significantly simpler and more straightforward to understand.

### CODE REVIEW

File: Userland/Applications/Piano/AudioPlayerLoop.cpp [hunk 1]
```diff
@@ -6,18 +6,94 @@
...
+
+        if (ns_duration_between(start_time, last_check_time) >= IPC::RESPONSIVENESS_TIMEOUT_MS * 1'000'000LL) {
+            timer_expired();
+            stop();
+        }
+    }
+
+    static long ns_duration_between(timespec start, timespec end)
+    {
+        auto s_duration = static_cast<i64>(end.tv_sec - start.tv_sec);
```
Comment (idispatch) [inline Userland/Applications/Piano/AudioPlayerLoop.cpp:47]: Shold we add a full second for when `end.tv_nsec - start.tv_nsec < 0` ?

```cpp
  if ((stop->tv_nsec - start->tv_nsec) < 0) {
        result->tv_sec = stop->tv_sec - start->tv_sec - 1;
        result->tv_nsec = stop->tv_nsec - start->tv_nsec + 1000000000;
    } else {
        result->tv_sec = stop->tv_sec - start->tv_sec;
        result->tv_nsec = stop->tv_nsec - start->tv_nsec;
    }
    ```

File: Userland/Applications/Piano/AudioPlayerLoop.h [hunk 2]
```diff
@@ -15,31 +15,41 @@
...
 // This does not run on a separate thread, preventing IPC multithreading madness.
 class AudioPlayerLoop final : public Core::Object {
     C_OBJECT(AudioPlayerLoop)
 public:
+    virtual ~AudioPlayerLoop() override;
+
     void enqueue_audio();
 
     void toggle_paused();
     bool is_playing() { return m_should_play_audio; }
```
Comment (idispatch) [inline Userland/Applications/Piano/AudioPlayerLoop.h:33]: `bool is_playing() const`

File: Userland/Applications/Piano/AudioPlayerLoop.cpp (hunk 1 above)
Comment (ADKaster) [reply Userland/Applications/Piano/AudioPlayerLoop.cpp:47]: Or... we could just use AK/Time.h and its AK::Time class, and all the fun helpers in that class, e.g. 

<https://github.com/SerenityOS/serenity/blob/master/AK/Time.h#L270-L279>

File: Userland/Libraries/LibAudio/Resampler.h [hunk 3]
```diff
@@ -66,6 +66,12 @@ class ResampleHelper {
         return resampled;
     }
 
+    template<ArrayLike<SampleType> Samples>
+    Vector<SampleType> resample(Samples&& to_resample)
+    {
+        return MUST(try_resample(to_resample));
```
Comment (idispatch) [inline Userland/Libraries/LibAudio/Resampler.h:72]: should `forward` be used here, since the function takes an rvalue reference?

**Review by ADKaster — CHANGES_REQUESTED** [review]
I have some comments on the code as you have it written here. 

However, I'm not convinced that your solution overall is correct.

What is the actual problem we are trying to solve here? "reduce issues due to shenanigans' doesn't really explain it. 

It feels like this is papering over some more fundamental architectural issue in AudioServer/AudioClient. In my thinking, I would expect the DAW application here that we are still calling Piano to work is to have a couple things going on:

- UI thread to process events from the user and push back events about what the audio they are hearing is doing to whatever visualizations or displays they have going on are.
- Audio thread doing mixing of samples to prepare to send them to the audio server for global mxiing
- Some processing to actually do the job of pushing data to the AudioServer.

Why does the thread that is mixing samples also send those samples to the audio server? Why does it matter if the thread pushing data to AudioServer uses an EventLoop? I know we discussed this on discord a bunch, but I still don't understand the goal or the rationale for using an alternative way to poke LibIPC in this way for just one application. 

Going a bit further, if there's something fancy that needs to happen with LibAudio's connection to AudioServer in order for it to work properly, that fix should go into the ConectionToServer class of LibAudio, not buried in the internals of one using application. Piano/DAW will not be the only application using audio that needs to work asynchronously.

File: Userland/Applications/Piano/AudioPlayerLoop.cpp (hunk 1 above)
Comment (kleinesfilmroellchen) [reply Userland/Applications/Piano/AudioPlayerLoop.cpp:47]: I totally missed that AK::Time exists.

File: Userland/Applications/Piano/AudioPlayerLoop.h (hunk 2 above)
Comment (kleinesfilmroellchen) [reply Userland/Applications/Piano/AudioPlayerLoop.h:33]: That's unrelated code :^)

File: Userland/Applications/Piano/AudioPlayerLoop.cpp [hunk 4]
```diff
@@ -28,37 +104,92 @@ AudioPlayerLoop::AudioPlayerLoop(TrackManager& track_manager, bool& need_to_writ
...
+    m_audio_client->shutdown();
 }
 
-void AudioPlayerLoop::timer_event(Core::TimerEvent&)
+intptr_t AudioPlayerLoop::pipeline_thread_main()
 {
-    while (m_audio_client->remaining_samples() < sample_count)
-        enqueue_audio();
+    m_audio_timer = &static_cast<AudioLoopResponsivenessTimer&>(m_audio_client->set_responsiveness_timer(make<AudioLoopResponsivenessTimer>(*m_audio_client)));
+    auto& deferred_invoker = static_cast<AudioLoopDeferredInvoker&>(m_audio_client->set_deferred_invoker(make<AudioLoopDeferredInvoker>()));
```
[the remark this answers is not in the document]
Comment (kleinesfilmroellchen) [reply Userland/Applications/Piano/AudioPlayerLoop.cpp:125]: Nope, this fix is specifically for when you don't use the convenience audio enqueueing APIs and don't have an event loop running.

File: Userland/Applications/Piano/AudioPlayerLoop.cpp [hunk 5]
```diff
@@ -28,37 +104,92 @@ AudioPlayerLoop::AudioPlayerLoop(TrackManager& track_manager, bool& need_to_writ
...
+    m_pipeline_thread->start();
+}
+
+AudioPlayerLoop::~AudioPlayerLoop()
+{
+    // Tell the pipeline to exit and wait for the last audio cycle to finish.
+    m_exit_requested.store(true);
+    auto result = m_pipeline_thread->join();
+    // FIXME: Get rid of the EINVAL/ESRCH check once we allow to join dead threads.
+    VERIFY(!result.is_error() || result.error() == EINVAL || result.error() == ESRCH);
```
[the remark this answers is not in the document]
Comment (kleinesfilmroellchen) [reply Userland/Applications/Piano/AudioPlayerLoop.cpp:117]: Maybe I did forget starting the thread and we would like to catch that.

File: Userland/Applications/Piano/AudioPlayerLoop.cpp [hunk 6]
```diff
@@ -6,18 +6,94 @@
...
+    {
+        deferred_functions.append(move(function));
+    }
+
+    void run_functions()
+    {
+        if (deferred_functions.size() > INLINE_FUNCTIONS)
+            dbgln("Warning: Audio loop has more than {} deferred functions, audio might glitch!", INLINE_FUNCTIONS);
+        if (!deferred_functions.is_empty())
+            dbgln("running {} functions", deferred_functions.size());
```
[the remark this answers is not in the document]
Comment (kleinesfilmroellchen) [reply Userland/Applications/Piano/AudioPlayerLoop.cpp:80]: It can go away, I wanted to see if there were even any deferred functions and it turns out that there are almost none in AudioServer communication.

File: Userland/Applications/Piano/AudioPlayerLoop.cpp [hunk 7]
```diff
@@ -28,37 +104,92 @@ AudioPlayerLoop::AudioPlayerLoop(TrackManager& track_manager, bool& need_to_writ
...
+
+    auto sample_rate = static_cast<double>(m_resampler->target());
+    auto buffer_play_time_ns = 1'000'000'000.0 / (sample_rate / static_cast<double>(Audio::AUDIO_BUFFER_SIZE));
+    auto good_sleep_time = Time::from_nanoseconds(static_cast<unsigned>(buffer_play_time_ns)).to_timespec();
+
+    size_t start_of_chunk_to_write = 0;
+    while (start_of_chunk_to_write + Audio::AUDIO_BUFFER_SIZE <= m_remaining_samples.size()) {
+        auto const exact_chunk = m_remaining_samples.span().slice(start_of_chunk_to_write, Audio::AUDIO_BUFFER_SIZE);
+        auto exact_chunk_array = Array<Audio::Sample, Audio::AUDIO_BUFFER_SIZE> {};
+        AK::TypedTransfer<Audio::Sample>::copy(exact_chunk_array.data(), exact_chunk.data(), Audio::AUDIO_BUFFER_SIZE);
```
[the remark this answers is not in the document]
Comment (kleinesfilmroellchen) [reply Userland/Applications/Piano/AudioPlayerLoop.cpp:160]: I did not find the nice API. In stage 2 I plan to introduce a small Array <- FixedArray API for this purpose.

File: Userland/Applications/Piano/AudioPlayerLoop.h [hunk 8]
```diff
@@ -15,31 +15,41 @@
...
-    virtual void timer_event(Core::TimerEvent&) override;
+    intptr_t pipeline_thread_main();
+    ErrorOr<void> send_audio_to_server();
+    void write_wav_if_needed();
 
     TrackManager& m_track_manager;
     FixedArray<DSP::Sample> m_buffer;
     Optional<Audio::ResampleHelper<DSP::Sample>> m_resampler;
     RefPtr<Audio::ConnectionToServer> m_audio_client;
+    AudioLoopResponsivenessTimer* m_audio_timer;
```
[the remark this answers is not in the document]
Comment (kleinesfilmroellchen) [reply Userland/Applications/Piano/AudioPlayerLoop.h:46]: The Connection owns this object. It is a raw pointer because it is only initialized once the audio pipeline thread starts (yes I would like it to be a reference, bad luck)

File: Userland/Applications/Piano/AudioPlayerLoop.cpp [hunk 9]
```diff
@@ -28,37 +104,92 @@ AudioPlayerLoop::AudioPlayerLoop(TrackManager& track_manager, bool& need_to_writ
...
-        m_track_manager.reset();
-        m_track_manager.set_should_loop(false);
-        do {
-            m_track_manager.fill_buffer(m_buffer);
-            m_wav_writer.write_samples(m_buffer.span());
-        } while (m_track_manager.transport()->time());
-        m_track_manager.reset();
-        m_track_manager.set_should_loop(true);
-        m_wav_writer.finalize();
+    TRY(m_remaining_samples.try_extend(TRY(m_resampler->try_resample(m_buffer))));
```
[the remark this answers is not in the document]
Comment (kleinesfilmroellchen) [reply Userland/Applications/Piano/AudioPlayerLoop.cpp:150]: (more APIs in this large-as-is PR)

File: Userland/Applications/Piano/AudioPlayerLoop.cpp [hunk 10]
```diff
@@ -28,37 +104,92 @@ AudioPlayerLoop::AudioPlayerLoop(TrackManager& track_manager, bool& need_to_writ
...
+        m_wav_writer.with_locked([this](auto& wav_writer) {
+            m_track_manager.reset();
+            m_track_manager.set_should_loop(false);
+            do {
+                m_audio_timer->update_time();
+                m_track_manager.fill_buffer(m_buffer);
+                wav_writer.write_samples(m_buffer.span());
+            } while (m_track_manager.transport()->time());
+            m_track_manager.reset();
+            m_track_manager.set_should_loop(true);
```
[the remark this answers is not in the document]
Comment (kleinesfilmroellchen) [reply Userland/Applications/Piano/AudioPlayerLoop.cpp:189]: This is copied code from the old WAV writing handler. It will be simplified once I move TrackManager and/or once rendering is handled better.

File: Userland/Applications/Piano/AudioPlayerLoop.cpp [hunk 11]
```diff
@@ -28,37 +104,92 @@ AudioPlayerLoop::AudioPlayerLoop(TrackManager& track_manager, bool& need_to_writ
...
+    }
+    m_remaining_samples.remove(0, start_of_chunk_to_write);
+    VERIFY(m_remaining_samples.size() < Audio::AUDIO_BUFFER_SIZE);
+
+    return {};
+}
+
+void AudioPlayerLoop::write_wav_if_needed()
+{
+    bool _true = true;
```
[the remark this answers is not in the document]
Comment (kleinesfilmroellchen) [reply Userland/Applications/Piano/AudioPlayerLoop.cpp:177]: (and you can't have a reference to a temporary)

File: Userland/Applications/Piano/main.cpp [hunk 12]
```diff
@@ -23,18 +24,21 @@
...
 #include <LibGUI/Window.h>
 #include <LibMain/Main.h>
+#include <LibThreading/MutexProtected.h>
 
 ErrorOr<int> serenity_main(Main::Arguments arguments)
 {
-    TRY(Core::System::pledge("stdio thread rpath cpath wpath recvfd sendfd unix"));
+    TRY(Core::System::pledge("stdio thread proc rpath cpath wpath recvfd sendfd unix"));
+
+    Core::EventLoop main_loop;
```
[the remark this answers is not in the document]
Comment (kleinesfilmroellchen) [reply Userland/Applications/Piano/main.cpp:33]: hmmm... I think there was a bug related to an event loop missing, as Application didn't create its loop soon enough.

File: Userland/Libraries/LibIPC/Connection.h [hunk 13]
```diff
@@ -27,21 +28,42 @@
 
 namespace IPC {
 
-// NOTE: This is an abstraction to allow using IPC::Connection without a Core::EventLoop.
+class ConnectionBase;
+
+// NOTE: These are abstractions to allow using IPC::Connection without a Core::EventLoop.
 // FIXME: It's not particularly nice, think of something nicer.
```
[the remark this answers is not in the document]
Comment (kleinesfilmroellchen) [reply Userland/Libraries/LibIPC/Connection.h:34]: I can, however that will mean breaking Ladybird.

File: Userland/Libraries/LibAudio/ConnectionToServer.cpp [hunk 14]
```diff
@@ -113,6 +114,11 @@ ErrorOr<void, AudioQueue::QueueStatus> ConnectionToServer::realtime_enqueue(Arra
     return m_buffer->try_enqueue(samples);
 }
 
+ErrorOr<void> ConnectionToServer::blocking_realtime_enqueue(Array<Sample, AUDIO_BUFFER_SIZE> samples, Function<void()> wait_function)
```
[the remark this answers is not in the document]
Comment (kleinesfilmroellchen) [reply Userland/Libraries/LibAudio/ConnectionToServer.cpp:117]: It's to clarify that this method has certain timing guarantees which the async enqueue functions don't have.

**Comment by kleinesfilmroellchen** [comment]
@ADKaster If you actually want all of these changes, I basically have to add 4-5 more large commits that add fancy APIs and refactor other parts of the code base. The only way of getting that sort of thing merged in a reasonable amount of time is to split it up into several PRs that add APIs without users and have no value in isolation.

**Comment by kleinesfilmroellchen** [comment]
Setting to draft because the IPC changes somehow immediately break all smooth audio playback. I have to investigate.

File: Userland/Applications/Piano/AudioPlayerLoop.h [hunk 15]
```diff
@@ -15,31 +15,41 @@
 #include <LibCore/Event.h>
 #include <LibCore/Object.h>
 #include <LibDSP/Music.h>
+#include <LibThreading/Thread.h>
 
 class TrackManager;
+struct AudioLoopResponsivenessTimer;
```
Comment (ADKaster) [inline Userland/Applications/Piano/AudioPlayerLoop.h:21]: this struct and the member are no longer used.

File: Userland/Libraries/LibIPC/Connection.h [hunk 16]
```diff
@@ -42,6 +42,7 @@ class ConnectionBase : public Core::Object {
 
     void set_fd_passing_socket(NonnullOwnPtr<Core::Stream::LocalSocket>);
     void set_deferred_invoker(NonnullOwnPtr<DeferredInvoker>);
+    DeferredInvoker& deferred_invoker() { return *m_deferred_invoker; }
```
Comment (ADKaster) [inline Userland/Libraries/LibIPC/Connection.h:45]: Any reason for this not to be `const` ?

**Review by ADKaster — CHANGES_REQUESTED** [review]
This solution is more palatable. But the request on discord for more numbers supporting changes that claim to improve performance still applies to future changes. Let's get some number driven audio fixes in rather than relying on each developer's ears. 

I did notice a handful of rebase artifacts. Please make sure to scroll through the commits on the github viewer to look for any unnecessary bits that are no longer necessary.

File: Userland/Libraries/LibIPC/Connection.h (hunk 16 above)
Comment (kleinesfilmroellchen) [reply Userland/Libraries/LibIPC/Connection.h:45]: I need to call non-const functions on the deferred invoker

**Review by ADKaster — APPROVED** [review]
All right, let's see how this goes. Thanks for working with me on this review.
